src/lexer.rs

```diff
@@ -0,0 +1,432 @@
+    CaretEqual,
+    Caret,
+    LeftBrace,
+    RightBrace,
+    Tilde,
+    PipeEqual,
+    DoublePipe,
+    Pipe,
+}
+
+#[derive(Debug)]
+pub struct Span {
+    pub start: Position,
+    pub end: Position,
+}
+
+#[derive(Debug)]
+pub struct Token<'t> {
+    pub span: Span,
+    pub data: TokenData<'t>,
+}
+
+impl<'t> Token<'t> {
+    fn new(start: Position, end: Position, value: TokenData<'t>) -> Self {
+        Token {
+            span: Span { start, end },
+            data: value,
+        }
+    }
+}
+
+pub struct Lexer<'t, I>
+where
+    I: Iterator<Item = char>,
+{
+    input: PositionedChars<I>,
+    strtab: &'t StringTable,
+    eof: bool,
+}
+
+impl<'t, I> Lexer<'t, I>
+where
+    I: Iterator<Item = char>,
+{
+    pub fn new(input: PositionedChars<I>, strtab: &'t StringTable) -> Self {
+        Lexer {
+            input,
+            strtab,
+            eof: false,
+        }
+    }
+
+    fn lex_token(&mut self) -> Token<'t> {
+        match self.input.peek() {
+            Some('a'..='z') | Some('A'..='Z') | Some('_') => self.lex_identifier_or_keyword(),
+
+            Some('0'..='9') => self.lex_integer_literal(),
+
+            Some(c) if c.is_whitespace() => self.lex_whitespace(),
+
+            Some(_) if self.input.peek_multiple(2) == "/*" => self.lex_comment(),
+
+            Some(_) => self.lex_operator().unwrap_or_else(|| {
+                let PositionedChar(pos, c) = self.input.next().unwrap();
+                Token::new(pos, pos, TokenData::UnexpectedCharacter(c))
+            }),
+
+            None => {
+                let pos = self.input.eof_position();
+                Token::new(pos, pos, TokenData::EOF)
+            }
+        }
+    }
+
+    fn lex_identifier_or_keyword(&mut self) -> Token<'t> {
+        self.lex_while(
+            |c| c.is_alphanumeric() || c == '_',
+            |ident, strtab, _| match Keyword::try_from(ident.as_ref()) {
+                Ok(keyword) => TokenData::Keyword(keyword),
+                Err(_) => TokenData::Identifier(strtab.intern(ident)),
+            },
+        )
+    }
+
+    fn lex_integer_literal(&mut self) -> Token<'t> {
+        self.lex_while(
+            |c| c.is_numeric(),
+            |lit, strtab, _| TokenData::IntegerLiteral(strtab.intern(lit)),
+        )
+    }
+
+    fn lex_comment(&mut self) -> Token<'t> {
+        self.lex_while_multiple(
+            2,
+            |s| s != "*/",
+            |text, _, eof_reached| {
+                if eof_reached {
+                    TokenData::UnclosedComment(text)
+                } else {
+                    TokenData::Comment(text)
+                }
+            },
+        )
+    }
+
+    fn lex_whitespace(&mut self) -> Token<'t> {
+        self.lex_while(|c| c.is_whitespace(), |_, _, _| TokenData::Whitespace)
+    }
+
+    #[allow(clippy::cyclomatic_complexity)]
+    fn lex_operator(&mut self) -> Option<Token<'t>> {
+        use self::Operator::*;
+
+        match self.input.peek_multiple(4) {
```

Comment:
Warum nicht andersherum testen? Zuerst das erste Zeichen testen und bei Uneindeutigkeit weiter das nächste Zeichen testen anstatt von vorne anzufangen?

Reply:
Ich habe gedacht, dass das dann so ein hässlich verzweigter Baum wird. Zugegebenermaßen ist das jetzt auch nicht viel schöner, aber zumindest kann man bei jedem Token auf den ersten Blick die zugehörige Zeichenkette erkennen, bei dem anderen müsste man immer den match-baum zurückverfolgen um das herauszufinden

Reply:
Darum -> Generator :D

Reply:
I like this version. It's not pretty, but it's easy to understand and hopefully Just Works™

Reply:
It will, however, get even uglier because implementing `Display` for `Keyword` and `Opetaror` will require to basically reimplement `Keyword::TryFrom<&str>` and `lex_operator` in reverese. Not very DRY...